src/main/python/deep_qa/layers/noisy_or.py

```diff
@@ -0,0 +1,93 @@
+        p = p + leaky_zeros_mask
+        # Clip values greater to 1 to 1
+        p *= K.cast(p <= 1., K.floatx())
+        leaky_ones_mask = K.cast(K.equal(p, K.zeros_like(p)), K.floatx())
+        p += leaky_ones_mask
+        return p
+
+
+class NoisyOr(Layer):
+    r"""
+    This layer takes as input a tensor of probabilities and calculates the noisy-or probability
+    across a given axis based on the noisy-or equation:
+        :math:`p(x) = 1 - \prod_{i=1:N}(1 - q * p(x|y_n))`
+    where q is the noise parameter
+
+    Inputs:
+        - probabilities, shape ``(batch, ..., N, ...)``,
+          where N is the number of y's in the above equation
+          (i.e. the number of probabilities being combined in the product),
+          in the dimension corresponding to the specified axis
+    Output:
+        - X, shape ``(batch, ..., ...)``, one less than input dims,
+          with the lost dimension corresponding to the specified axis.
+
+    Parameters
+    ----------
+    axis : int, default=-1
+        The axis over which to combine probabilities.
+    name : string, default='noisy_or'
+        Name of the layer, ued to debug both the layer and its parameter.
+    param_init : string, default='uniform'
+        The initialization of the noise parameter.
+    noise_param_constraint : Keras Constraint, default=None
+        Optional, a constraint which would be applied to the noise parameter.
+    """
+    def __init__(self, axis=-1, name="noisy_or", param_init='uniform', noise_param_constraint=None, **kwargs):
+        self.axis = axis
+        self.param_init = param_init
+        self.supports_masking = True
+        self.name = name
+        # The noisy-or equation includes a noise parameter (q) which is learned during training
+        self.noise_parameter = None
+        self.noise_param_constraint = noise_param_constraint
+        super(NoisyOr, self).__init__(**kwargs)
+
+    def build(self, input_shape):
+        # Add the trainable weight variable for the noise parameter
+        self.noise_parameter = self.add_weight(shape=(),
+                                               name=self.name + '_noise_param',
+                                               initializer=self.param_init,
+                                               regularizer=L1L2Regularizer(l2=0.001),
+                                               constraint=self.noise_param_constraint,
+                                               trainable=True)
+        super(NoisyOr, self).build(input_shape)
+
+    def get_output_shape_for(self, input_shape):
+        if self.axis == -1:
+            return input_shape[0:-1]
+        return input_shape[:self.axis - 1] + input_shape[self.axis:]
+
+    @overrides
+    def compute_mask(self, inputs, mask=None):
+        # pylint: disable=unused-argument
+        if mask is not None:
+            return K.any(mask, axis=-1)
+        return None
+
+    def call(self, inputs, mask=None):
+        # dim: (b, ..., num_probs, ...)
```

Comment:
We've typically formatted shape comments either like [this](https://github.com/allenai/deep_qa/blob/ba567499184b27e6f7c1747360ab86b54c5e8277/src/main/python/deep_qa/models/multiple_choice_qa/decomposable_attention.py#L49), or like [this](https://github.com/allenai/deep_qa/blob/ba567499184b27e6f7c1747360ab86b54c5e8277/src/main/python/deep_qa/models/reading_comprehension/attention_sum_reader.py#L71).  Because the comment is just the shape, no period necessary here, but I would try to follow one of those two styles.  Either use `shape` or nothing (instead of `dim`), and expand `b` to `batch_size` or `batch size`.